packages/backend-common/src/util/docker.ts

```diff
@@ -0,0 +1,127 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Docker from 'dockerode';
+import { PassThrough, Writable } from 'stream';
+
+export type UserOptions = {
+  User?: string;
+};
+
+export type RunDockerContainerOptions = {
+  imageName: string;
+  args: string[];
+  logStream?: Writable;
+  dockerClient: Docker;
+  mountDirs?: Map<string, string>;
+  workingDir?: string;
+  envVars?: string[];
+  createOptions?: Docker.ContainerCreateOptions;
+};
+
+/**
+ *
+ * @param options the options object
+ * @param options.imageName the image to run
+ * @param options.args the arguments to pass the container
+ * @param options.logStream the log streamer to capture log messages
+ * @param options.dockerClient the dockerClient to use
+ * @param options.mountDirs A map of host directories to mount on the container.
+ *        Map Key: Path on host machine, Value: Path on Docker container
+ * @param options.workingDir Working dir in the container
+ * @param options.envVars Environment variables to set in the container. e.g. ['HOME=/tmp']
+ */
+export const runDockerContainer = async ({
+  imageName,
+  args,
+  logStream = new PassThrough(),
+  dockerClient,
+  mountDirs = new Map(),
+  workingDir,
+  envVars = [],
+  createOptions = {},
+}: RunDockerContainerOptions) => {
+  // Show a better error message when Docker is unavailable.
+  try {
+    await dockerClient.ping();
+  } catch (e) {
+    throw new Error(
+      `This operation requires Docker. Docker does not appear to be available. Docker.ping() failed with: ${e.message}`,
+    );
+  }
+
+  await new Promise<void>((resolve, reject) => {
+    dockerClient.pull(imageName, {}, (err, stream) => {
+      if (err) return reject(err);
+      stream.pipe(logStream, { end: false });
+      stream.on('end', () => resolve());
+      stream.on('error', (error: Error) => reject(error));
+      return undefined;
+    });
+  });
+
+  const userOptions: UserOptions = {};
+  if (process.getuid && process.getgid) {
+    // Files that are created inside the Docker container will be owned by
+    // root on the host system on non Mac systems, because of reasons. Mainly the fact that
+    // volume sharing is done using NFS on Mac and actual mounts in Linux world.
+    // So we set the user in the container as the same user and group id as the host.
+    // On Windows we don't have process.getuid nor process.getgid
+    userOptions.User = `${process.getuid()}:${process.getgid()}`;
+  }
+
+  // Initialize volumes to mount based on mountDirs map
+  const Volumes: { [T: string]: object } = {};
+  for (const containerDir of mountDirs.values()) {
+    Volumes[containerDir] = {};
+  }
+
+  // Create bind volumes
+  const Binds: string[] = [];
+  for (const [hostDir, containerDir] of mountDirs.entries()) {
+    Binds.push(`${hostDir}:${containerDir}`);
```

Comment:
Could call `realpath` on the host dir here instead of having the caller do it tbh. The overhead of that is pretty minimal compared to other things going on, and it's pretty easy to miss as a user of the API.